File: specs/typespec/admin/tags.ts
import {
	type ValidationError,
	newValidationError,
	ERR_REQUIRED,
} from "../common/common";

// Tag ID: lowercase letters and hyphens, no leading/trailing hyphens, 1-64 chars
const TAG_ID_PATTERN = /^[a-z]([a-z0-9-]*[a-z0-9])?$/;
const TAG_ID_MAX_LENGTH = 64;

export const ERR_TAG_ID_REQUIRED = "tag_id is required";
export const ERR_TAG_ID_INVALID_FORMAT =
	"tag_id must contain only lowercase letters, digits, and hyphens, and must not start or end with a hyphen";
export const ERR_TAG_ID_TOO_LONG = `tag_id must be at most ${TAG_ID_MAX_LENGTH} characters`;
export const ERR_TRANSLATIONS_REQUIRED = "at least one translation is required";
export const ERR_EN_US_TRANSLATION_REQUIRED = "en-US translation is required";
export const ERR_DISPLAY_NAME_REQUIRED = "display_name is required";
export const ERR_DISPLAY_NAME_TOO_LONG =
	"display_name must be at most 100 characters";
export const ERR_DESCRIPTION_TOO_LONG =
	"description must be at most 500 characters";
export const ERR_LOCALE_REQUIRED = "locale is required";
export const ERR_ICON_SIZE_INVALID = "icon_size must be 'small' or 'large'";

export type TagId = string;
export type IconSize = "small" | "large";

export interface TagTranslation {
	locale: string;
	display_name: string;
	description?: string;
}

export function validateTagTranslation(
	translation: TagTranslation,
	index: number
): ValidationError[] {
	const errs: ValidationError[] = [];
	const prefix = `translations[${index}]`;

	if (!translation.locale || translation.locale.trim() === "") {
		errs.push(newValidationError(`${prefix}.locale`, ERR_LOCALE_REQUIRED));
	}

	if (!translation.display_name || translation.display_name.trim() === "") {
		errs.push(
			newValidationError(`${prefix}.display_name`, ERR_DISPLAY_NAME_REQUIRED)
		);
	} else if (translation.display_name.length > 100) {
		errs.push(
			newValidationError(`${prefix}.display_name`, ERR_DISPLAY_NAME_TOO_LONG)
		);
	}

	if (
		translation.description !== undefined &&
		translation.description !== null
	) {
		if (translation.description.length > 500) {
			errs.push(
				newValidationError(`${prefix}.description`, ERR_DESCRIPTION_TOO_LONG)
			);
		}
	}

	return errs;
}

export function validateTagId(tagId: TagId): string | null {
	if (!tagId || tagId.trim() === "") {
		return ERR_TAG_ID_REQUIRED;
	}
	if (tagId.length > TAG_ID_MAX_LENGTH) {
		return ERR_TAG_ID_TOO_LONG;
	}
	if (!TAG_ID_PATTERN.test(tagId)) {
		return ERR_TAG_ID_INVALID_FORMAT;
	}
	return null;
}

export interface CreateTagRequest {
	tag_id: TagId;
	translations: TagTranslation[];
}

export function validateCreateTagRequest(
	request: CreateTagRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	const tagIdErr = validateTagId(request.tag_id);
	if (tagIdErr) {
		errs.push(newValidationError("tag_id", tagIdErr));
	}

	if (!request.translations || request.translations.length === 0) {
		errs.push(newValidationError("translations", ERR_TRANSLATIONS_REQUIRED));
	} else {
		const hasEnUS = request.translations.some((t) => t.locale === "en-US");
		if (!hasEnUS) {
			errs.push(
				newValidationError("translations", ERR_EN_US_TRANSLATION_REQUIRED)
			);
		}
		request.translations.forEach((translation, index) => {
			const translationErrs = validateTagTranslation(translation, index);
			errs.push(...translationErrs);
		});
	}

	return errs;
}

export interface UpdateTagRequest {
	tag_id: TagId;
	translations: TagTranslation[];
}

export function validateUpdateTagRequest(
	request: UpdateTagRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	const tagIdErr = validateTagId(request.tag_id);
	if (tagIdErr) {
		errs.push(newValidationError("tag_id", tagIdErr));
	}

	if (!request.translations || request.translations.length === 0) {
		errs.push(newValidationError("translations", ERR_TRANSLATIONS_REQUIRED));
	} else {
		const hasEnUS = request.translations.some((t) => t.locale === "en-US");
		if (!hasEnUS) {
			errs.push(
				newValidationError("translations", ERR_EN_US_TRANSLATION_REQUIRED)
			);
		}
		request.translations.forEach((translation, index) => {
			const translationErrs = validateTagTranslation(translation, index);
			errs.push(...translationErrs);
		});
	}

	return errs;
}

export interface GetTagRequest {
	tag_id: TagId;
}

export function validateGetTagRequest(
	request: GetTagRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	const tagIdErr = validateTagId(request.tag_id);
	if (tagIdErr) {
		errs.push(newValidationError("tag_id", tagIdErr));
	}

	return errs;
}

export interface FilterTagsRequest {
	query?: string;
	pagination_key?: string;
}

export function validateFilterTagsRequest(
	_request: FilterTagsRequest
): ValidationError[] {
	return [];
}

export interface DeleteTagIconRequest {
	tag_id: TagId;
	icon_size: IconSize;
}

export function validateDeleteTagIconRequest(
	request: DeleteTagIconRequest
): ValidationError[] {
	const errs: ValidationError[] = [];

	const tagIdErr = validateTagId(request.tag_id);
	if (tagIdErr) {
		errs.push(newValidationError("tag_id", tagIdErr));
	}

	if (request.icon_size !== "small" && request.icon_size !== "large") {
		errs.push(newValidationError("icon_size", ERR_ICON_SIZE_INVALID));
	}

	return errs;
}

export interface AdminTag {
	tag_id: TagId;
	translations: TagTranslation[];
	small_icon_url?: string;
	large_icon_url?: string;
	created_at: string;
	updated_at: string;
}

export interface FilterTagsResponse {
	tags: AdminTag[];
	pagination_key?: string;
}

// Re-export ValidationError for use in consumers
export type { ValidationError };
export { ERR_REQUIRED };

File: specs/typespec/admin/tags.go
package admin

import (
	"fmt"
	"regexp"

	"vetchium-api-server.typespec/common"
)

var tagIDPattern = regexp.MustCompile(`^[a-z]([a-z0-9-]*[a-z0-9])?$`)

const tagIDMaxLength = 64

const (
	errTagIDRequired           = "tag_id is required"
	errTagIDTooLong            = "tag_id must be at most 64 characters"
	errTagIDInvalidFormat      = "tag_id must contain only lowercase letters, digits, and hyphens, and must not start or end with a hyphen"
	errTranslationsRequired    = "at least one translation is required"
	errEnUSTranslationRequired = "en-US translation is required"
	errDisplayNameRequired     = "display_name is required"
	errDisplayNameTooLong      = "display_name must be at most 100 characters"
	errDescriptionTooLong      = "description must be at most 500 characters"
	errLocaleRequired          = "locale is required"
	errIconSizeInvalid         = "icon_size must be 'small' or 'large'"
)

type IconSize string

const (
	IconSizeSmall IconSize = "small"
	IconSizeLarge IconSize = "large"
)

// TagTranslation holds a locale and its display name/description.
type TagTranslation struct {
	Locale      string  `json:"locale"`
	DisplayName string  `json:"display_name"`
	Description *string `json:"description,omitempty"`
}

// validateTagID validates a tag_id value and returns validation errors.
func validateTagID(tagID string) []common.ValidationError {
	var errs []common.ValidationError
	if tagID == "" {
		errs = append(errs, common.NewValidationError("tag_id", fmt.Errorf(errTagIDRequired)))
		return errs
	}
	if len(tagID) > tagIDMaxLength {
		errs = append(errs, common.NewValidationError("tag_id", fmt.Errorf(errTagIDTooLong)))
		return errs
	}
	if !tagIDPattern.MatchString(tagID) {
		errs = append(errs, common.NewValidationError("tag_id", fmt.Errorf(errTagIDInvalidFormat)))
	}
	return errs
}

// validateTranslations validates a slice of TagTranslation and returns validation errors.
func validateTranslations(translations []TagTranslation) []common.ValidationError {
	var errs []common.ValidationError
	if len(translations) == 0 {
		errs = append(errs, common.NewValidationError("translations", fmt.Errorf(errTranslationsRequired)))
		return errs
	}
	hasEnUS := false
	for i, t := range translations {
		prefix := fmt.Sprintf("translations[%d]", i)
		if t.Locale == "" {
			errs = append(errs, common.NewValidationError(prefix+".locale", fmt.Errorf(errLocaleRequired)))
		} else {
			if err := common.LanguageCode(t.Locale).Validate(); err != nil {
				errs = append(errs, common.NewValidationError(prefix+".locale", err))
			}
			if t.Locale == "en-US" {
				hasEnUS = true
			}
		}
		if t.DisplayName == "" {
			errs = append(errs, common.NewValidationError(prefix+".display_name", fmt.Errorf(errDisplayNameRequired)))
		} else if len(t.DisplayName) > 100 {
			errs = append(errs, common.NewValidationError(prefix+".display_name", fmt.Errorf(errDisplayNameTooLong)))
		}
		if t.Description != nil && len(*t.Description) > 500 {
			errs = append(errs, common.NewValidationError(prefix+".description", fmt.Errorf(errDescriptionTooLong)))
		}
	}
	if !hasEnUS {
		errs = append(errs, common.NewValidationError("translations", fmt.Errorf(errEnUSTranslationRequired)))
	}
	return errs
}

// CreateTagRequest is the request body for POST /admin/add-tag.
type CreateTagRequest struct {
	TagID        string           `json:"tag_id"`
	Translations []TagTranslation `json:"translations"`
}

func (r CreateTagRequest) Validate() []common.ValidationError {
	errs := validateTagID(r.TagID)
	errs = append(errs, validateTranslations(r.Translations)...)
	return errs
}

// UpdateTagRequest is the request body for POST /admin/update-tag.
type UpdateTagRequest struct {
	TagID        string           `json:"tag_id"`
	Translations []TagTranslation `json:"translations"`
}

func (r UpdateTagRequest) Validate() []common.ValidationError {
	errs := validateTagID(r.TagID)
	errs = append(errs, validateTranslations(r.Translations)...)
	return errs
}

// GetTagRequest is the request body for POST /admin/get-tag.
type GetTagRequest struct {
	TagID string `json:"tag_id"`
}

func (r GetTagRequest) Validate() []common.ValidationError {
	return validateTagID(r.TagID)
}

// FilterTagsRequest is the request body for POST /admin/filter-tags.
type FilterTagsAdminRequest struct {
	Query         string `json:"query,omitempty"`
	PaginationKey string `json:"pagination_key,omitempty"`
}

func (r FilterTagsAdminRequest) Validate() []common.ValidationError {
	return nil
}

// DeleteTagIconRequest is the request body for POST /admin/delete-tag-icon.
type DeleteTagIconRequest struct {
	TagID    string   `json:"tag_id"`
	IconSize IconSize `json:"icon_size"`
}

func (r DeleteTagIconRequest) Validate() []common.ValidationError {
	errs := validateTagID(r.TagID)
	if r.IconSize != IconSizeSmall && r.IconSize != IconSizeLarge {
		errs = append(errs, common.NewValidationError("icon_size", fmt.Errorf(errIconSizeInvalid)))
	}
	return errs
}

// AdminTag is the response type for tag operations.
type AdminTag struct {
	TagID        string           `json:"tag_id"`
	Translations []TagTranslation `json:"translations"`
	SmallIconURL *string          `json:"small_icon_url,omitempty"`
	LargeIconURL *string          `json:"large_icon_url,omitempty"`
	CreatedAt    string           `json:"created_at"`
	UpdatedAt    string           `json:"updated_at"`
}

// FilterTagsAdminResponse is the response for POST /admin/filter-tags.
type FilterTagsAdminResponse struct {
	Tags          []AdminTag `json:"tags"`
	PaginationKey string     `json:"pagination_key,omitempty"`
}
